src/filelock/_api.py

```diff
@@ -73,15 +75,36 @@ class ThreadLocalFileContext(FileLockContext, local):
     """A thread local version of the ``FileLockContext`` class."""
 
 
-class BaseFileLock(ABC, contextlib.ContextDecorator):
+class SingletonPerLockFileMeta(ABCMeta):
+    """
+    Metaclass for the ``BaseFileLock`` class which ensures that only one instance of the class is created per
+    lock (when specified).
+    """
+
+    _instances: ClassVar[WeakValueDictionary] = WeakValueDictionary()
+
+    def __call__(cls, lock_file: str | os.PathLike[str], *args: Sequence, **kwargs: Mapping) -> object:
+        if not kwargs.get("singleton_per_lock_file", False):
+            return super().__call__(lock_file, *args, **kwargs)
+
+        instance = cls._instances.get(str(lock_file))
+        if not instance:
+            instance = super().__call__(lock_file, *args, **kwargs)
+            cls._instances[str(lock_file)] = instance
+
+        return instance
+
+
+class BaseFileLock(ABC, contextlib.ContextDecorator, metaclass=SingletonPerLockFileMeta):
     """Abstract base class for a file lock object."""
 
-    def __init__(
+    def __init__(  # noqa: PLR0913
         self,
         lock_file: str | os.PathLike[str],
         timeout: float = -1,
         mode: int = 0o644,
         thread_local: bool = True,  # noqa: FBT001, FBT002
+        singleton_per_lock_file: bool = False,  # noqa: FBT001, FBT002
```

Comment:
Naming suggestions welcome.

Reply:
Make it keyword only.

Reply:
Perhaps just `singleton` should suffice.